lang/java/reef-runtime-azbatch/src/main/java/org/apache/reef/runime/azbatch/driver/AzureBatchResourceManager.java

```diff
@@ -45,54 +45,89 @@
 
   private final Map<String, ResourceRequestEvent> containerRequests;
-
+  private final Set<String> activeContainerIds;
   private final ConfigurationSerializer configurationSerializer;
   private final AzureBatchEvaluatorShimManager evaluatorShimManager;
+  private final AzureBatchTaskStatusAlarmHandler azureBatchTaskStatusAlarmHandler;
 
   private final double jvmHeapFactor;
   private final CommandBuilder launchCommandBuilder;
+  private final AtomicInteger containerCount;
 
   @Inject
   AzureBatchResourceManager(
       final ConfigurationSerializer configurationSerializer,
       final CommandBuilder launchCommandBuilder,
       final AzureBatchEvaluatorShimManager evaluatorShimManager,
+      final AzureBatchTaskStatusAlarmHandler azureBatchTaskStatusAlarmHandler,
       @Parameter(JVMHeapSlack.class) final double jvmHeapSlack) {
     this.configurationSerializer = configurationSerializer;
     this.evaluatorShimManager = evaluatorShimManager;
     this.jvmHeapFactor = 1.0 - jvmHeapSlack;
     this.launchCommandBuilder = launchCommandBuilder;
     this.containerRequests = new ConcurrentHashMap<>();
+    this.activeContainerIds = Collections.synchronizedSet(new HashSet<String>());
+    this.containerCount = new AtomicInteger(0);
+    this.azureBatchTaskStatusAlarmHandler = azureBatchTaskStatusAlarmHandler;
   }
 
   public void onResourceRequested(final ResourceRequestEvent resourceRequestEvent) {
     LOG.log(Level.FINEST, "Got ResourceRequestEvent in AzureBatchResourceManager,");
     for (int r = 0; r < resourceRequestEvent.getResourceCount(); r++) {
       final String containerId = generateContainerId();
       this.containerRequests.put(containerId, resourceRequestEvent);
+      this.containerCount.incrementAndGet();
+      this.azureBatchTaskStatusAlarmHandler.enableAlarm();
       this.evaluatorShimManager.onResourceRequested(containerId, resourceRequestEvent);
     }
   }
 
   public void onResourceReleased(final ResourceReleaseEvent resourceReleaseEvent) {
-    LOG.log(Level.FINEST, "Got ResourceReleasedEvent for Id: {0} in AzureBatchResourceManager",
-        resourceReleaseEvent.getIdentifier());
-    this.containerRequests.remove(resourceReleaseEvent.getIdentifier());
+    String id = resourceReleaseEvent.getIdentifier();
+    LOG.log(Level.FINEST, "Got ResourceReleasedEvent for Id: {0} in AzureBatchResourceManager", id);
+    if (!this.activeContainerIds.remove(id)) {
+      LOG.log(Level.WARNING,
+          "Attempting to remove non-existent activeContainer for Id: {0} in AzureBatchResourceManager", id);
+    }
+
+    ResourceRequestEvent removedEvent = this.containerRequests.remove(id);
+    if (removedEvent == null) {
+      LOG.log(Level.WARNING,
+          "Ignoring attempt to remove non-existent containerRequest for Id: {0} in AzureBatchResourceManager", id);
+    }
+    int currentContainerCount = this.containerCount.decrementAndGet();
```

Comment:
So if this function is triggered for the the same containerId multiple times, currentContainerCount will get decremented for each invocation. Which will make the count negative. You may want to move it into an 'else' block for one of the 'if' statements above.

Reply:
done.

Reply:
Reed makes a good point. We are creating quite a few tracking items; activeContainerIds, containerRequests, containerCount. I have to wonder if all these are all necessary.